pkg/auth/providers/azure/azure_provider.go

```diff
@@ -422,105 +333,55 @@ func (ap *azureProvider) getAzureConfigK8s() (*v32.AzureADConfig, error) {
-			if err != nil {
-				logrus.Debugf("[AZURE_PROVIDER] Error getting group: %v", err)
-				return err
-			}
-
-			p := ap.groupToPrincipal(groupObj)
-			p.MemberOf = true
-
-			// Add to the cache
-			groupCache.Add(gp, p)
-			groupPrincipals[j] = p
-			return nil
-		})
-	}
-	if err := g.Wait(); err != nil {
-		return nil, err
-	}
-	logrus.Debugf("[AZURE_PROVIDER] Completed gathering users groups, took %v, keys in cache:%v", time.Since(start), groupCache.Len())
-	return groupPrincipals, nil
-}
-
-func (ap *azureProvider) groupToPrincipal(group graphrbac.ADGroup) v3.Principal {
-	p := v3.Principal{
-		ObjectMeta:    metav1.ObjectMeta{Name: Name + "_group://" + *group.ObjectID},
-		DisplayName:   *group.DisplayName,
-		PrincipalType: "group",
-		Provider:      Name,
-	}
-	return p
-}
-
-func (ap *azureProvider) getAzureToken(secret string) (adal.Token, error) {
-	var azureToken adal.Token
-	err := json.Unmarshal([]byte(secret), &azureToken)
+// updateToken compares the current Azure access token to the one stored in the secret and updates if needed.
+// This is relevant only for access tokens to the deprecated Azure AD Graph API.
+func (ap *azureProvider) updateToken(client clients.AzureClient, token *v3.Token) error {
+	// For the new flow via Microsoft Graph, the caching and updating of the token to the Microsoft Graph API
+	// is handled separately via the SDK client cache.
+	cfg, err := ap.getAzureConfigK8s()
 	if err != nil {
-		return azureToken, err
+		return err
+	}
+	if !graphEndpointDeprecated(cfg) {
+		return nil
 	}
-	return azureToken, nil
-}
 
-// updateToken compares the current azure token to the azure token living in the
-// secret and updates if needed
-func (ap *azureProvider) updateToken(client *azureClient, token *v3.Token) error {
-	new, err := client.marshalTokenJSON()
+	current, err := client.MarshalTokenJSON()
 	if err != nil {
-		return err
+		return errors.New("failed to unmarshal token")
 	}
-	stringNew := string(new)
 
 	secret, err := ap.tokenMGR.GetSecret(token.UserID, token.AuthProvider, []*v3.Token{token})
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			// providerToken doesn't exists as a secret, update on token
-			if current, ok := token.ProviderInfo["access_token"]; ok && current != stringNew {
-				token.ProviderInfo["access_token"] = stringNew
+			// providerToken doesn't exist as a secret, update on token.
+			if current, ok := token.ProviderInfo["access_token"]; ok && current != current {
+				token.ProviderInfo["access_token"] = current
 			}
 			return nil
 		}
 		return err
 	}
 
-	if stringNew == secret {
+	if current == secret {
 		return nil
 	}
 
-	return ap.tokenMGR.UpdateSecret(token.UserID, token.AuthProvider, stringNew)
+	return ap.tokenMGR.UpdateSecret(token.UserID, token.AuthProvider, current)
 }
 
 func formAzureRedirectURL(config map[string]interface{}) string {
+	authEndpoint, ok := config["authEndpoint"].(string)
```

Comment:
Rather than trying to parse this out of the map (which is tricky and could lead to issues since authEndpoint is being used as the source of truth for deprecation here, but not in the case of the the commonly used method), you could use `mapstructure.decode()` to turn this into a well structured AzureConfig object and then use the `graphEndpointDeprecated` function, which is more consistent with our approaches elsewhere. 

The one thing that I am unsure about would be the danger of trying to convert here - it might be that the configmap isn't a valid authconfig. But I'm not 100% sure on that.

```suggestion
	azureConfig := &v32.AzureADConfig{}
	err := mapstructure.Decode(config, azureConfig)
	if err == nil && !graphEndpointDeprecated(azureConfig) {
		return fmt.Sprintf(
			"%s?client_id=%s&redirect_uri=%s&response_type=code&scope=user.read",
			azureConfig.AuthEndpoint,
			azureConfig.ApplicationID,
			azureConfig.RancherURL,
		)
	}
	return fmt.Sprintf(
		"%s?client_id=%s&redirect_uri=%s&resource=%s",
		config["authEndpoint"],
		config["applicationId"],
		config["rancherUrl"],
		config["graphEndpoint"],
	)
```